pkg/route/validation/validation_test.go

```diff
@@ -843,6 +981,522 @@ func TestValidateRoute(t *testing.T) {
+			headerType:           "response",
+		},
+		{
+			name: "Should give an error if the response header's value has sample converter req.hdr",
+			route: &routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "wildcardpolicy",
+					Namespace: "foo",
+				},
+				Spec: routev1.RouteSpec{
+					Host:           "subdomain.wildcard.test",
+					To:             createRouteSpecTo("serviceName", "Service"),
+					WildcardPolicy: routev1.WildcardPolicySubdomain,
+					HTTPHeaders: &routev1.RouteHTTPHeaders{
+
+						Actions: routev1.RouteHTTPHeaderActions{
+							Response: []routev1.RouteHTTPHeader{
+								{
+									Name: "X-SSL-Client-Cert",
+									Action: routev1.RouteHTTPHeaderActionUnion{
+										Type: routev1.Set,
+										Set: &routev1.RouteSetHTTPHeader{
+											Value: "%[req.hdr(host),lower]",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedErrorMessage: "spec.httpHeaders.actions.response[0].action.set.value: Invalid value: \"X-SSL-Client-Cert\": Either header value provided is not in correct format or the converter specified is not allowed. The dynamic header value  may use HAProxy's %[] syntax and otherwise must be a valid HTTP header value as defined in https://datatracker.ietf.org/doc/html/rfc7230#section-3.2 Sample fetchers allowed are res.hdr, ssl_c_der. Converters allowed are lower, base64.",
+		},
+		{
+			name: "should not allow repetition of a header name",
+			route: &routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "wildcardpolicy",
+					Namespace: "foo",
+				},
+				Spec: routev1.RouteSpec{
+					Host:           "subdomain.wildcard.test",
+					To:             createRouteSpecTo("serviceName", "Service"),
+					WildcardPolicy: routev1.WildcardPolicySubdomain,
+					HTTPHeaders: &routev1.RouteHTTPHeaders{
+
+						Actions: routev1.RouteHTTPHeaderActions{
+							Request: []routev1.RouteHTTPHeader{
+								{
+									Name: "Accept",
+									Action: routev1.RouteHTTPHeaderActionUnion{
+										Type: routev1.Set,
+										Set: &routev1.RouteSetHTTPHeader{
+											Value: "text/plain,text/html",
+										},
+									},
+								},
+								{
+									Name: "Accept",
+									Action: routev1.RouteHTTPHeaderActionUnion{
+										Type: routev1.Delete,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedErrorMessage: "spec.httpHeaders.actions.request[1].name: Invalid value: \"Accept\": header name has been repeated",
+			headerType:           "request",
+		},
+	}
+	for _, tc := range tests {
+		var errorMessage *field.Error
+		switch tc.headerType {
+		case "response":
+			errorMessage = validateHeaders(field.NewPath(specResponseString), tc.route.Spec.HTTPHeaders.Actions.Response)
+		case "request":
+			errorMessage = validateHeaders(field.NewPath(specRequestString), tc.route.Spec.HTTPHeaders.Actions.Request)
+		}
+		if errorMessage != nil && errorMessage.Error() != "" {
```

Comment:
You still have to check whether we expected an empty error message, if we received one.  I think that's why some of these tests are passing, but should be failing.  Remove this condition.

Reply:
we want to check errors only if they are available or else goroutines will crash nil error.